src/rgw/driver/rados/topic.cc

```diff
@@ -0,0 +1,383 @@
+
+  bufferlist bl;
+  cache_entry entry;
+  rgw_cache_entry_info cache_info;
+  int r = rgw_get_system_obj(&sysobj, obj.pool, obj.oid, bl, &entry.objv,
+                             &entry.mtime, y, dpp, nullptr, &cache_info);
+  if (r < 0) {
+    return r;
+  }
+
+  try {
+    auto p = bl.cbegin();
+    decode(entry.info, p);
+  } catch (const buffer::error&) {
+    return -EIO;
+  }
+
+  cache.put(dpp, &cache_svc, topic_key, &entry, {&cache_info});
+
+  if (pmtime) {
+    *pmtime = entry.mtime;
+  }
+  if (pobjv) {
+    *pobjv = std::move(entry.objv);
+  }
+  info = std::move(entry.info);
+  return 0;
+}
+
+int write(const DoutPrefixProvider* dpp, optional_yield y,
+          RGWSI_SysObj& sysobj, RGWSI_MDLog* mdlog, const RGWZoneParams& zone,
+          const rgw_pubsub_topic& info, RGWObjVersionTracker& objv,
+          ceph::real_time mtime, bool exclusive)
+{
+  const std::string topic_key = get_topic_metadata_key(info.user.tenant, info.name);
+  const rgw_raw_obj obj = get_topic_obj(zone, topic_key);
+
+  bufferlist bl;
+  encode(info, bl);
+
+  int r = rgw_put_system_obj(dpp, &sysobj, obj.pool, obj.oid,
+                             bl, exclusive, &objv, mtime, y);
+  if (r < 0) {
+    ldpp_dout(dpp, 1) << "ERROR: failed to write topic obj " << obj.oid
+        << " with: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  // record in the mdlog on success
+  if (mdlog) {
+    return mdlog->complete_entry(dpp, y, "topic", topic_key, &objv);
+  }
+  return 0;
+}
+
+int remove(const DoutPrefixProvider* dpp, optional_yield y,
+           RGWSI_SysObj& sysobj, RGWSI_MDLog* mdlog, const RGWZoneParams& zone,
+           const std::string& topic_key, RGWObjVersionTracker& objv)
+{
+  // delete topic info
+  const rgw_raw_obj topic = get_topic_obj(zone, topic_key);
+  int r = rgw_delete_system_obj(dpp, &sysobj, topic.pool, topic.oid, &objv, y);
+  if (r < 0) {
+    ldpp_dout(dpp, 1) << "ERROR: failed to remove topic obj "
+        << topic.oid << " with: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  // delete the buckets object
+  const rgw_raw_obj buckets = get_buckets_obj(zone, topic_key);
+  r = rgw_delete_system_obj(dpp, &sysobj, buckets.pool,
+                            buckets.oid, nullptr, y);
+  if (r < 0) {
+    ldpp_dout(dpp, 20) << "WARNING: failed to remove topic buckets obj "
+        << buckets.oid << " with: " << cpp_strerror(r) << dendl;
+  } // not fatal
+
+  // record in the mdlog on success
+  if (mdlog) {
+    return mdlog->complete_entry(dpp, y, "topic", topic_key, &objv);
+  }
+  return 0;
+}
+
+
+int link_bucket(const DoutPrefixProvider* dpp, optional_yield y,
+                librados::Rados& rados, const RGWZoneParams& zone,
+                const std::string& topic_key,
+                const std::string& bucket_key)
+{
+  const rgw_raw_obj obj = get_buckets_obj(zone, topic_key);
+
+  rgw_rados_ref ref;
+  int r = rgw_get_rados_ref(dpp, &rados, obj, &ref);
```

Comment:
In previous case we when the `ioctx` that was initialized by calling `rgw_init_ioctx`, the param passed was `mostly_omap=true`, and now when `ioctx` is initialized by calling `rgw_get_rados_ref` it passes `mostly_omap=false`.
looking at the definition of `rgw_init_ioctx()` for `omap=true` its setting `pg_autoscale_bias` n `recovery_priority`. Not sure if that has any implications?

Reply:
thanks, i'd missed that

regarding `mostly_omap`, i don't think we want that set for the topic pool. we do use omap for the topic-bucket mapping objects, but we'll expect at least as many data objects for the topics themselves. and because we're namespaced under the same pool as other metadata by default, we should use the same pool flags for all of them

ioctx is a light-weight handle to a rados pool, so it's okay to create one for each librados op. sharing a global ioctx instance in RGWRados doesn't save us much. and if the underlying pool gets deleted/recreated, that global handle goes stale until rgw restarts

on RGWRados startup, it might make sense to ensure that all the common pools get created. but in a cluster that doesn't use notifications, this topic pool may not ever need to be created. so i think it's better to create it on demand